### PR TITLE
[core] Allow user to override global default for max_retries

### LINKED ISSUE #24854: [core] Allow user to configure default max task retries
### Description

The default for the `ray.remote(max_retries)` parameter is hardcoded as a Ray constant. It can be overridden for individual functions or task calls, but this is a bit awkward if you want to change the max retries for a specific library or application. The library has to expose it and make sure to reset the max_retries for every remote function that it defines.

We should improve this by allowing the user to override the default. One way is to set a global default for the application with an environment variable. A more advanced version is to have nested defaults to better support applications that mix different libraries. For example, one library can set its max_retries default, and all child tasks will inherit the same max_retries.

### Use case

Example: Allowing users to set higher task retries for Datasets to improve fault tolerance.

For nested defaults, a use case would be the above + another part of the application where reconstruction is not desired.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This PR allows the user to override the global default for max_retries for non-actor tasks. It adds an OS env called `RAY_task_max_retries` which can be passed to the driver or set with runtime envs. Any future tasks submitted by that worker will default to this value instead of 3, the hard-coded default.

It would be nicer if we could have a standard way of setting these defaults, but I think this is fine as a one-off for now (not a clear need for overriding defaults of other @ray.remote options yet).

## Related issue number

Closes #24854.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/remote_function.py [hunk 1]
```diff
@@ -217,6 +218,15 @@ def _remote(self, args=None, kwargs=None, **task_options):
...
         for k, v in ray_option_utils.task_options.items():
+            if k == "max_retries":
+                # TODO(swang): We need to override max_retries here because the default
+                # value gets set at Ray import time. Ideally, we should allow setting
+                # default values from env vars for other options too.
+                v.default_value = os.environ.get(
+                    "RAY_task_max_retries", v.default_value
+                )
+                v.default_value = int(v.default_value)
+                print("option", k, v.default_value)
```
Comment (scv119) [inline python/ray/remote_function.py:229]: looks like debug log?

Comment (stephanie-wang) [reply python/ray/remote_function.py]: Ah, thanks! :|

File: doc/source/ray-core/tasks/fault-tolerance.rst [hunk 2]
```diff
@@ -8,7 +8,10 @@ because the process crashed or because the machine failed, Ray will rerun
 the task until either the task succeeds or the maximum number of retries is
 exceeded. The default number of retries is 3 and can be overridden by
 specifying ``max_retries`` in the ``@ray.remote`` decorator. Specifying -1
-allows infinite retries, and 0 disables retries.
+allows infinite retries, and 0 disables retries. To override the default number
+of retries for all tasks submitted, set the OS environment variable
+``RAY_task_max_retries``. e.g., by passing this to your driver script or by
+using :ref:`runtime environments<runtime-environments>`.
```
Comment (ericl) [inline doc/source/ray-core/tasks/fault-tolerance.rst:14]: By convention, I think we should add an alias so this is `RAY_TASK_MAX_RETRIES`, and reserve `RAY_lower_case_vars` as internal flags, for user clarity.

I'll leave it to you and @scv119 to decide if this should be a policy though.

File: python/ray/remote_function.py [hunk 3]
```diff
@@ -217,6 +218,14 @@ def _remote(self, args=None, kwargs=None, **task_options):
 
         # fill task required options
         for k, v in ray_option_utils.task_options.items():
+            if k == "max_retries":
+                # TODO(swang): We need to override max_retries here because the default
```
Comment (scv119) [inline python/ray/remote_function.py:222]: ah this is tricky... wondering if there is a better way.

**Review by scv119 — APPROVED** [review]

File: doc/source/ray-core/tasks/fault-tolerance.rst (hunk 2 above)
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/fault-tolerance.rst:14]: Makes sense, I'll change this.

File: doc/source/ray-core/tasks/fault-tolerance.rst [hunk 4]
```diff
@@ -8,7 +8,10 @@ because the process crashed or because the machine failed, Ray will rerun
 the task until either the task succeeds or the maximum number of retries is
 exceeded. The default number of retries is 3 and can be overridden by
 specifying ``max_retries`` in the ``@ray.remote`` decorator. Specifying -1
-allows infinite retries, and 0 disables retries.
+allows infinite retries, and 0 disables retries. To override the default number
+of retries for all tasks submitted, set the OS environment variable
+``RAY_TASK_MAX_RETRIES``. e.g., by passing this to your driver script or by
+using :ref:`runtime environments<runtime-environments>`.
```
Comment (rkooo567) [inline doc/source/ray-core/tasks/fault-tolerance.rst:14]: Why don't we have a short example here? 
```
RAY_TASK_MAX_RETRIES driver.py

or 

# in driver.py
ray.init(address="auto", runtime_env={...})
```

File: python/ray/remote_function.py (hunk 3 above)
Comment (rkooo567) [reply python/ray/remote_function.py:222]: Just curious, but Is this not going to work?
```
    "max_retries": _counting_option(
        "max_retries", default_value=os.environ.get("RAY_TASK_MAX_RETRIES", ray_constants.DEFAULT_TASK_MAX_RETRIES)
    ),
```
?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
cc @matthewdeng

File: python/ray/remote_function.py (hunk 3 above)
Comment (stephanie-wang) [reply python/ray/remote_function.py:222]: It doesn't work if the OS env is set after the ray import. Maybe it's only a real issue for python tests, but either way checking it at ray.remote time seems more robust.

File: doc/source/ray-core/tasks/fault-tolerance.rst (hunk 4 above)
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/fault-tolerance.rst:14]: I'm not sure this is necessary (runtime envs page provides examples) and it kind of breaks the flow of this docs page.

**Pull request merged**